src/raftstore/store/upload.rs

```diff
@@ -0,0 +1,390 @@
+
+pub struct UploadFile {
+    meta: SSTMeta,
+    save_path: PathBuf,
+    temp_path: PathBuf,
+    temp_file: Option<File>,
+    temp_digest: crc32::Digest,
+}
+
+impl UploadFile {
+    fn create(meta: SSTMeta, save_path: PathBuf, temp_path: PathBuf) -> Result<UploadFile> {
+        let temp_file = File::create(&temp_path)?;
+        Ok(UploadFile {
+            meta: meta,
+            save_path: save_path,
+            temp_path: temp_path,
+            temp_file: Some(temp_file),
+            temp_digest: crc32::Digest::new(crc32::IEEE),
+        })
+    }
+
+    fn append(&mut self, data: &[u8]) -> Result<()> {
+        self.temp_file.as_mut().unwrap().write_all(data)?;
+        self.temp_digest.write(data);
+        Ok(())
+    }
+
+    fn finish(&mut self) -> Result<()> {
+        self.validate()?;
+        self.temp_file.take();
+        assert!(self.temp_path.exists());
+        assert!(!self.save_path.exists());
+        fs::rename(&self.temp_path, &self.save_path)
+    }
+
+    fn remove(&mut self) -> Result<()> {
+        self.temp_file.take();
+        if self.temp_path.exists() {
+            fs::remove_file(&self.temp_path)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn validate(&self) -> Result<()> {
+        let f = self.temp_file.as_ref().unwrap();
+        if f.metadata()?.len() != self.meta.get_len() {
+            return Err(file_corrupted(&self.temp_path));
+        }
+        if self.temp_digest.sum32() != self.meta.get_crc32() {
+            return Err(file_corrupted(&self.temp_path));
+        }
+        Ok(())
+    }
+}
+
+impl Drop for UploadFile {
+    fn drop(&mut self) {
+        if let Err(e) = self.remove() {
+            warn!("remove {}: {:?}", self, e);
+        }
+    }
+}
+
+impl Display for UploadFile {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "UploadFile {{meta: {{{:?}}}, save_path: {:?}, temp_path: {:?}}}",
+            self.meta,
+            self.save_path,
+            self.temp_path
+        )
+    }
+}
+
+fn sst_handle_to_path(h: &SSTHandle) -> Result<PathBuf> {
+    let uuid = match Uuid::from_bytes(h.get_uuid()) {
+        Ok(uuid) => uuid,
+        Err(e) => {
+            let error = format!("invalid uuid {:?}: {:?}", h.get_uuid(), e);
+            return Err(Error::new(ErrorKind::InvalidInput, error));
+        }
+    };
+
+    if h.get_cfname().is_empty() || h.get_region_id() == 0 ||
+        h.get_region_epoch().get_conf_ver() == 0 || h.get_region_epoch().get_version() == 0
+    {
+        let error = format!("invalid sst handle {:?}", h);
+        return Err(Error::new(ErrorKind::InvalidInput, error));
+    }
+
+    Ok(PathBuf::from(format!(
+        "{}_{}_{}_{}_{}.sst",
+        uuid.simple().to_string(),
+        h.get_cfname(),
+        h.get_region_id(),
+        h.get_region_epoch().get_conf_ver(),
+        h.get_region_epoch().get_version(),
+    )))
+}
+
+fn file_exists<P: AsRef<Path>>(path: P) -> Error {
```

Comment:
seem we should move these common functions to util mod.

Reply:
my fault, these functions are for error, I think we should name `file_exists_error`